// standardized-audio-context/src/types/audio-listener-factory-factory.ts
import { TAudioListenerFactory } from 'standardized-audio-context/src/types/audio-listener-factory';
import { TAudioParamFactory } from 'standardized-audio-context/src/types/audio-param-factory';
import { TIsNativeOfflineAudioContextFunction } from 'standardized-audio-context/src/types/is-native-offline-audio-context-function';
import { TNativeChannelMergerNodeFactory } from 'standardized-audio-context/src/types/native-channel-merger-node-factory';
import { TNativeConstantSourceNodeFactory } from 'standardized-audio-context/src/types/native-constant-source-node-factory';
import { TNativeScriptProcessorNodeFactory } from 'standardized-audio-context/src/types/native-script-processor-node-factory';

export type TAudioListenerFactoryFactory = (
    createAudioParam: TAudioParamFactory,
    createNativeChannelMergerNode: TNativeChannelMergerNodeFactory,
    createNativeConstantSourceNode: TNativeConstantSourceNodeFactory,
    createNativeScriptProcessorNode: TNativeScriptProcessorNodeFactory,
    isNativeOfflineAudioContext: TIsNativeOfflineAudioContextFunction
) => TAudioListenerFactory;
