// standardized-audio-context/src/types/offline-audio-context-constructor-factory.ts
import { IOfflineAudioContextConstructor } from 'standardized-audio-context/src/interfaces/index';
import { TBaseAudioContextConstructor } from 'standardized-audio-context/src/types/base-audio-context-constructor';
import { TCacheTestResultFunction } from 'standardized-audio-context/src/types/cache-test-result-function';
import { TCreateNativeOfflineAudioContextFunction } from 'standardized-audio-context/src/types/create-native-offline-audio-context-function';
import { TInvalidStateErrorFactory } from 'standardized-audio-context/src/types/invalid-state-error-factory';
import { TStartRenderingFunction } from 'standardized-audio-context/src/types/start-rendering-function';

export type TOfflineAudioContextConstructorFactory = (
    baseAudioContextConstructor: TBaseAudioContextConstructor,
    cacheTestResult: TCacheTestResultFunction,
    createInvalidStateError: TInvalidStateErrorFactory,
    createNativeOfflineAudioContext: TCreateNativeOfflineAudioContextFunction,
    startRendering: TStartRenderingFunction
) => IOfflineAudioContextConstructor;
