// tone/Tone/fromContext.ts
import * as Classes from "tone/Tone/classes";
import { Transport } from "tone/Tone/core/clock/Transport";
import { Context } from "tone/Tone/core/context/Context";
import { Listener } from "tone/Tone/core/context/Listener";
import { Destination } from "tone/Tone/core/context/Destination";
import { FrequencyClass } from "tone/Tone/core/type/Frequency";
import { MidiClass } from "tone/Tone/core/type/Midi";
import { TicksClass } from "tone/Tone/core/type/Ticks";
import { TimeClass } from "tone/Tone/core/type/Time";
import { TransportTimeClass } from "tone/Tone/core/type/TransportTime";
import { isDefined, isFunction } from "tone/Tone/core/util/TypeCheck";
import { omitFromObject } from "tone/Tone/core/util/Defaults";
import { Draw } from "tone/Tone/core/util/Draw";

type ClassesWithoutSingletons = Omit<typeof Classes, "Transport" | "Destination" | "Draw">;

/**
 * The exported Tone object. Contains all of the classes that default
 * to the same context and contains a singleton Transport and Destination node.
 */
type Tone = {
	Transport: Transport;
	Destination: Destination;
	Listener: Listener;
	Draw: Draw;
	context: Context;
	now: () => number;
	immediate: () => number;
} & ClassesWithoutSingletons;

/**
 * Bind the TimeBaseClass to the context
 */
function bindTypeClass(context: Context, type) {
	return (...args: any[]) => new type(context, ...args);
}

/**
 * Return an object with all of the classes bound to the passed in context
 * @param context The context to bind all of the nodes to
 */
export function fromContext(context: Context): Tone {

	const classesWithContext: Partial<ClassesWithoutSingletons> = {};
	Object.keys(omitFromObject(Classes, ["Transport", "Destination", "Draw"])).map(key => {
		const cls = Classes[key];
		if (isDefined(cls) && isFunction(cls.getDefaults)) {
			classesWithContext[key] = class ToneFromContextNode extends cls {
				get defaultContext(): Context {
					return context;
				}
			};
		} else {
			// otherwise just copy it over
			classesWithContext[key] = Classes[key];
		}
	});

	const toneFromContext: Tone = {
		...(classesWithContext as ClassesWithoutSingletons),
		now: context.now.bind(context),
		immediate: context.immediate.bind(context),
		Transport: context.transport,
		Destination: context.destination,
		Listener: context.listener,
		Draw: context.draw,
		context,
		// the type functions
		Midi: bindTypeClass(context, MidiClass),
		Time: bindTypeClass(context, TimeClass),
		Frequency: bindTypeClass(context, FrequencyClass),
		Ticks: bindTypeClass(context, TicksClass),
		TransportTime: bindTypeClass(context, TransportTimeClass),
	};
	// return the object
	return toneFromContext;
}
