// standardized-audio-context/src/interfaces/minimal-base-audio-context.ts
import { TAudioContextState, TContext, TEventHandler, TNativeEventTarget } from 'standardized-audio-context/src/types/index';
import { IAudioDestinationNode } from 'standardized-audio-context/src/interfaces/audio-destination-node';
import { IAudioListener } from 'standardized-audio-context/src/interfaces/audio-listener';
import { IMinimalBaseAudioContextEventMap } from 'standardized-audio-context/src/interfaces/minimal-base-audio-context-event-map';

export interface IMinimalBaseAudioContext<T extends TContext> extends TNativeEventTarget {
    readonly currentTime: number;

    readonly destination: IAudioDestinationNode<T>;

    readonly listener: IAudioListener;

    onstatechange: null | TEventHandler<T>;

    readonly sampleRate: number;

    readonly state: TAudioContextState;

    addEventListener<K extends keyof IMinimalBaseAudioContextEventMap>(
        type: K,
        listener: (this: this, event: IMinimalBaseAudioContextEventMap[K]) => void,
        options?: boolean | AddEventListenerOptions
    ): void;

    addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void;

    removeEventListener<K extends keyof IMinimalBaseAudioContextEventMap>(
        type: K,
        listener: (this: this, event: IMinimalBaseAudioContextEventMap[K]) => void,
        options?: boolean | EventListenerOptions
    ): void;

    removeEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | EventListenerOptions): void;
}
